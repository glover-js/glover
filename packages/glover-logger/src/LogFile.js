// @flow
import type { LogMessage } from 'types/Log';
import type { Path } from 'types/Path';

import os from 'os';
import fs from 'fs';
import path from 'path';

export default class LogFile {
  _filename: string;
  _dir: Path;
  _outputPath: Path;
  _outputStream: fs.WriteStream;

  constructor(outputPath: ?string) {
    this._filename = 'glover-debug.log';
    this._dir = process.cwd();
    this._outputPath = outputPath || path.join(this._dir, this._filename);
    this._outputStream = fs.createWriteStream(this._outputPath);
    this._outputStream.setDefaultEncoding('utf8');
    this._outputStream.on('error', err => this._handleError(err));
    this._ensureStreamIsWritable();
  }

  write(message: LogMessage): void {
    const timeString = new Date().toISOString();
    const lineBreak = os.EOL || '\n';
    const formattedMessage = `${timeString}: ${message} ${lineBreak}`;
    this._outputStream.write(formattedMessage);
  }

  close(): Promise<void> {
    return new Promise(resolve => {
      this._outputStream.end(resolve);
    });
  }

  _handleError(err: Error & { code?: string }): void {
    const code = err.code ? `(${err.code})` : '';
    const error: any = new Error();
    error.code = err.code;
    error.message = `Could not write log to: ${this._outputPath} (${code})`;
    throw error;
  }

  _ensureStreamIsWritable(): void {
    // Synchronously check if we will be able to write to this._outputPath
    try {
      fs.accessSync(this._outputPath, fs.constants.W_OK);
    } catch (err) {
      this._handleError(err);
    }
  }
}
