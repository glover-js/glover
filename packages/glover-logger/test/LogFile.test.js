import os from 'os';
import fs from 'fs';
import LogFile from '../src/LogFile';

jest.mock('fs', () => {
  const MemoryFs = require('memory-fs'); // eslint-disable-line global-require
  const memoryFs = new MemoryFs();
  memoryFs.constants = { W_OK: 1 };
  memoryFs.accessSync = () => true;
  return memoryFs;
});

describe('Tests for LogFile', () => {
  describe('constructor', () => {
    it('throws an error when providing a invalid output path', () => {
      const accessSync = jest.spyOn(fs, 'accessSync');
      accessSync.mockImplementation(() => {
        throw new Error(); // Fake that the path cannot be accessed
      });
      const outputPath = '/not-existing-path/test.log';
      expect(() => new LogFile(outputPath)).toThrowErrorMatchingSnapshot();
      accessSync.mockRestore();
    });
  });
  describe('.write', () => {
    it('writes the given message to the logfile', () => {
      const EOL = os.EOL || '\n';
      const toISOString = jest.spyOn(Date.prototype, 'toISOString');
      toISOString.mockImplementation(() => '[DATE]');
      const outputDir = '/log';
      const outputPath = '/log/test.log';
      fs.mkdirpSync(outputDir); // Create the output directory
      const logFile = new LogFile(outputPath);
      logFile.write('[TEST] message');
      logFile.close();
      const actual = fs.readFileSync(outputPath);
      expect(actual.toString('utf8')).toEqual(`[DATE]: [TEST] message ${EOL}`);
    });
  });
});
