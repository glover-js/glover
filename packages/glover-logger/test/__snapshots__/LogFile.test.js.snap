// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for LogFile constructor throws an error when providing a invalid output path 1`] = `"Could not write log to: /not-existing-path/test.log ()"`;
